Guide GCP Cloud Trace with OpenTelemetry (#964)
micronaut-projects/micronaut-guides: guides/micronaut-cloud-trace-google/micronaut-cloud-trace-google.adoc

external:micronaut-cloud-trace-base/start.adoc[]
* A https://cloud.google.com/gcp/[Google Cloud Platform (GCP)] account and a GCP project.

common:costs-gcloud.adoc[]

common:gcp.adoc[]

external:micronaut-cloud-trace-base/create-app.adoc[]

=== Tracer Configuration

> Using Micronaut CLI or Launch to create your application, the OpenTelemetry exporter will be added to your `application.yml` configurations.

Reviewer: ```suggestion
Use Micronaut CLI or Launch to create your application. The OpenTelemetry exporter will be automatically added to your `application.yml` configurations.
```

resource:application.yml[tag=otel-trace]

== Run the Application

> The application can be deployed to Google Cloud Run, Cloud Function, Cloud Compute or App Engine.

Reviewer: ```suggestion
The application can be deployed to Google Cloud Run, Cloud Function, Cloud Compute, or App Engine.
```

Traces can be sent to Google Cloud Trace outside the Google Cloud.

This allows us to run the application locally and see the traces in Google Cloud Trace.

common:runapp-instructions.adoc[]

> [NOTE]

Reviewer: Same comment from the other guide; should there be a note here telling the reader that this error is nothing to worry about?

====
You might get this error message: +
"Message: The Application Default Credentials are not available. They are available if running in Google Compute Engine. Otherwise, the environment variable GOOGLE_APPLICATION_CREDENTIALS must be defined pointing to a file defining the credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information."

Login with developer credentials:
[,bash]
----
gcloud auth application-default login
----
====

== Traces

Open the https://console.cloud.google.com/traces[GCP Tracing Console].

image::gcp-tracing/trace-list-graph.png[]

{empty} +

> It does take a few seconds for the traces to show up in the console.

Reviewer: ```suggestion
If the data is not displayed, give it a minute. It takes a few seconds for the traces to show up in the console.
```

=== Get Item Counts

[source, bash]
----
curl http://localhost:8080/store/inventory/laptop
----
{empty} +

image::gcp-tracing/inventory-item.png[]

{empty} +

Each span is represented by a blue bar.

=== Order Item

[source, bash]
----
curl -X "POST" "http://localhost:8080/store/order" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{"item":"laptop", "count":5}'
----

{empty} +

image::gcp-tracing/order-item.png[]

{empty} +

> Selecting a different span will show you the labels (a.k.a attributes/tags) and other details of the span.

Reviewer: ```suggestion
Selecting a different span will show you the labels (a.k.a. attributes/tags) and other details of the span.
```

=== Get Inventory

[source, bash]
----
curl http://localhost:8080/store/inventory
----

{empty} +

image::gcp-tracing/inventory-all.png[]

{empty} +

> Looking at the trace, retrieving the items sequentially might not be the best design choice.

Reviewer: ```suggestion
Looking at the trace, we can conclude that retrieving the items sequentially might not be the best design choice.
```

external:micronaut-cloud-trace-base/start.adoc[]
* A https://cloud.google.com/gcp/[Google Cloud Platform (GCP)] account and a GCP project.

common:costs-gcloud.adoc[]

common:gcp.adoc[]

external:micronaut-cloud-trace-base/create-app.adoc[]

=== Configure Tracer

TODO: Fix when starter is updated

Add the following snippet to include the necessary dependencies:

.build.gradle
[,groovy]
----

> annotationProcessor("io.micronaut.tracing:micronaut-tracing-opentelemetry-annotation:4.2.0-SNAPSHOT")

Reviewer 2: I think this next two dependencies should be part of the base guide.

```
annotationProcessor("io.micronaut.tracing:micronaut-tracing-opentelemetry-annotation:4.2.0-SNAPSHOT")
implementation("io.micronaut.tracing:micronaut-tracing-opentelemetry:4.2.0-SNAPSHOT")
```
